### PR TITLE
 Updates the minimizing correlations tutorial for TensorFlow 2.0

### PR DESCRIPTION
**Pull request by josh146**
**Context:** The existing TensorFlow tutorial was out of date (requiring the old TF1.3 backend), and in the form of a Jupyter notebook.

**Description of the Change:**

* Ports the tutorial to Sphinx-Gallery, by using the [converter script available here](https://gist.github.com/chsasank/7218ca16f8d022e02a9c0deb94a310fe).

* Ports the tutorial to use the new TF2 backend

**Benefits:** As the tutorial has been ported to Sphinx gallery, it is now executable, making it easy to test to see whether it provides the correct output.

**Possible Drawbacks:** n/a

**Related GitHub Issues:** n/a

### CODE REVIEW

**Comment by josh146** [comment]
Note: as far as I can tell, this works perfectly with TF 2.0 installed (cost function minimizes to -0.9122365 after 200 steps), but on TF 2.1, cost function grows to 1.22043 by step 200 🙁

File: examples_apps/run_minimizing_correlations.py [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+# the initial state of mode ``q[1]``, which we will optimize over. Note
+# that we take the outer product
+# :math:`\ket{in}=\ket{\psi}\otimes\ket{\phi}`, and use the ``Ket``
+# operator to initialise the circuit in this initial multimode pure state.
+
+
+def cost(phi_var_re, phi_var_im):
+    phi_var = tf.cast(phi_var_re, tf.complex64) + 1j*tf.cast(phi_var_im, tf.complex64)
+
+    in_state = tf.einsum('i,j->ij', psi, phi_var)
```
Comment (trbromley) [inline examples_apps/run_minimizing_correlations.py:191]: Just a note that `einsum` is here.

File: examples_apps/run_minimizing_correlations.py [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+ket_val = phi_var_re.numpy() + phi_var_im.numpy()*1j
+out_rhoB = np.outer(ket_val, ket_val.conj())
+
+
+######################################################################
+# Next, we can use the following function to plot the Wigner function of
+# this density matrix:
+#
+
+import copy
```
Comment (nquesada) [inline examples_apps/run_minimizing_correlations.py:273]: I guess now that is a python script we can no longer just import qutip here, right?

File: examples_apps/run_minimizing_correlations.py [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+# quadrature and by the same amount.
+#
+# Now imagine the following task: > Given an input state
+# :math:`\left|\psi \right\rangle`, which is not necessarily a squeezed
+# state, what is the optimal state :math:`\left|\phi \right\rangle`
+# incident on a beamsplitter :math:`\text{BS}(\theta)` together with
+# :math:`\left|\psi \right\rangle` such that the resulting entanglement is
+# minimized?
+#
+# In our `paper <https://arxiv.org/abs/1805.06868>`__ we showed that if
```
Comment (nquesada) [inline examples_apps/run_minimizing_correlations.py:31]:
```suggestion
# In our `paper <https://journals.aps.org/pra/abstract/10.1103/PhysRevA.98.043813>`__ we showed that if
```

File: examples_apps/run_minimizing_correlations.py [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+from scipy.linalg import expm
+import tensorflow as tf
+
+import strawberryfields as sf
+from strawberryfields.ops import *
+from strawberryfields.backends.tfbackend.ops import partial_trace
+
+
+# set the random seed
+tf.random.set_seed(137)
```
Comment (nquesada) [inline examples_apps/run_minimizing_correlations.py:55]: <3

File: examples_apps/run_minimizing_correlations.py [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+# References
+# ----------
+#
+# [1] Jiang, Z., Lang, M. D., & Caves, C. M. (2013). Mixing nonclassical
+# pure states in a linear-optical network almost always generates modal
+# entanglement. *Physical Review A*, 88(4), 044301.
+#
+# [2] Quesada, N., & Brańczyk, A. M. (2018). Gaussian functions are
+# optimal for waveguided nonlinear-quantum-optical processes.
+# *arXiv:1805.06868*.
```
Comment (nquesada) [inline examples_apps/run_minimizing_correlations.py:354]:
```suggestion
# *Physical Review A*, 98, 043813.
```

File: examples_apps/run_minimizing_correlations.py (hunk 1 above)
Comment (josh146) [reply examples_apps/run_minimizing_correlations.py:191]: Oh nice point!

File: examples_apps/run_minimizing_correlations.py (hunk 2 above)
Comment (josh146) [reply examples_apps/run_minimizing_correlations.py:273]: Oh, the original jupyter notebook wasn't importing Qutip either? I didn't make any changes to this section

Comment (nquesada) [reply examples_apps/run_minimizing_correlations.py:273]: It did, but I was hoping it could be removed.

**Comment by josh146** [comment]
@nquesada one thing I might get you to look at --- for some reason, the trace of the state after optimization is a bit far below 1 (and as a result, the probability of state |0> is above 1).

**Comment by nquesada** [comment]
That is fine, the optimization is not guaranteed to keep the state normalized. Simply renormalize after the minimization happens and all should be OK.

File: examples_apps/run_minimizing_correlations.py (hunk 1 above)
Comment (trbromley) [reply examples_apps/run_minimizing_correlations.py:191]: https://github.com/XanaduAI/strawberryfields/pull/377

File: examples_apps/run_minimizing_correlations.py [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+#
+
+penalty_strength = 10
+
+prog = sf.Program(2)
+eng = sf.Engine("tf", backend_options={"cutoff_dim": cutoff})
+
+psi = tf.cast(psi, tf.complex64)
+phi_var_re = tf.Variable(phi.real)
+phi_var_im = tf.Variable(phi.imag)
```
Comment (thisac) [inline examples_apps/run_minimizing_correlations.py:171]: Are the imaginary and real parts separated just for clarity, or does are real variables somehow needed for the cost function below?

File: examples_apps/run_minimizing_correlations.py [hunk 7]
```diff
@@ -0,0 +1,355 @@
...
+    # Stores cost at each step
+    cost_progress.append(loss.numpy())
+
+    # one repetition of the optimization
+    gradients = tape.gradient(loss, [phi_var_re, phi_var_im])
+    opt.apply_gradients(zip(gradients, [phi_var_re, phi_var_im]))
+
+    # Prints progress
+    if i % 1 == 0:
+        print("Rep: {} Cost: {}".format(i, loss))
```
Comment (thisac) [inline examples_apps/run_minimizing_correlations.py:248]: No f-string? :wink:

File: examples_apps/run_minimizing_correlations.py (hunk 6 above)
Comment (josh146) [reply examples_apps/run_minimizing_correlations.py:171]: Unfortunately, TF2.0 doesn't support optimizing complex variables, so it had to be separated. TF2.1 adds support again, but then breaks complex gradient computations 😆

**Pull request merged**